projects/packages/account-protection/src/class-validation-service.php

```diff
@@ -0,0 +1,326 @@
+		return $length < Config::VALIDATION_SERVICE_MIN_LENGTH || $length > Config::VALIDATION_SERVICE_MAX_LENGTH;
+	}
+
+	/**
+	 * Check if the password matches any user data.
+	 *
+	 * @param \WP_User|\stdClass $user The user.
+	 * @param string             $password The password to check.
+	 *
+	 * @return bool True if the password matches any user data, false otherwise.
+	 */
+	public function matches_user_data( $user, string $password ): bool {
+		if ( ! $user ) {
+			return false;
+		}
+
+		$email_parts    = explode( '@', $user->user_email ); // test@example.com
+		$email_username = $email_parts[0]; // 'test'
+		$email_domain   = $email_parts[1]; // 'example.com'
+		$email_provider = explode( '.', $email_domain )[0]; // 'example'
+
+		$user_data = array(
+			$user->user_login ?? '',
+			$user->display_name ?? '',
+			$user->first_name ?? '',
+			$user->last_name ?? '',
+			$user->user_email ?? '',
+			$email_username ?? '',
+			$email_provider ?? '',
+			$user->nickname ?? '',
+		);
+
+		$password_lower = strtolower( $password );
+
+		foreach ( $user_data as $data ) {
+			// Skip if $data is 3 characters or less.
+			if ( strlen( $data ) <= 3 ) {
+				continue;
+			}
+
+			if ( ! empty( $data ) && strpos( $password_lower, strtolower( $data ) ) !== false ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Check if the password is in the list of compromised/common passwords.
+	 *
+	 * @param string $password The password to check.
+	 *
+	 * @return bool True if the password is in the list of compromised/common passwords, false otherwise.
+	 */
+	public function is_weak_password( string $password ): bool {
+		if ( ! $this->connection_manager->is_connected() ) {
+			return false;
+		}
+
+		$hashed_password = sha1( $password );
+		$password_prefix = substr( $hashed_password, 0, 5 );
+
+		$response = $this->request_suffixes( $password_prefix );
+
+		$response_code = wp_remote_retrieve_response_code( $response );
+
+		if ( is_wp_error( $response ) || 200 !== $response_code || empty( $response['body'] ) ) {
+			return false;
+		}
+
+		$body = json_decode( wp_remote_retrieve_body( $response ), true );
+
+		$password_suffix = substr( $hashed_password, 5 );
+		if ( in_array( $password_suffix, $body['compromised'] ?? array(), true ) ) {
+			return true;
+		}
+
+		if ( in_array( $password_suffix, $body['common'] ?? array(), true ) ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Check if the password is the current password for the user.
+	 *
+	 * @param int    $user_id  The user ID.
+	 * @param string $password The password to check.
+	 *
+	 * @return bool True if the password is the current password, false otherwise.
+	 */
+	public function is_current_password( int $user_id, string $password ): bool {
+		$user = get_userdata( $user_id );
+		if ( ! $user ) {
+			return false;
+		}
+
+		return wp_check_password( $password, $user->user_pass, $user->ID );
+	}
+
+	/**
+	 * Check if the password has been used recently by the user.
```

Comment:
I don't know if that's possible, but I think it would be nice if this check supported checking if you use a recent password with just an added letter for example, or a recent password with just one character changed? 

I know this is sometimes common practice in places where they force employees to update their passwords every few months: folks end up changing from `MyDogName1` to 'MyDogName2` when it's time to change.

Reply:
Added this request to our backlog for the next iteration!